lib/pulse/lib/Pulse.Lib.ArrayPtr.fsti

```diff
@@ -0,0 +1,159 @@
+module Pulse.Lib.ArrayPtr
+open FStar.Tactics.V2
+open Pulse.Lib.Pervasives
+module SZ = FStar.SizeT
+module A = Pulse.Lib.Array
+
+(* This module can be used only for Pulse programs extracted to C, not Rust. *)
+
+val ptr : Type0 -> Type0
+
+[@@erasable]
+val footprint : Type0 -> Type0
+
+val pts_to (#t: Type) (s: ptr t) (#[exact (`1.0R)] p: perm) (fp: footprint t) (v: Seq.seq t) : slprop
+
+val pts_to_is_slprop2 (#a:Type) (x:ptr a) (p:perm) (fp: footprint a) (s:Seq.seq a)
+  : Lemma (is_slprop2 (pts_to x #p fp s))
+          [SMTPat (is_slprop2 (pts_to x #p fp s))]
+
+val is_from_array (#t: Type) (p: perm) (fp: footprint t) (a: A.array t) : slprop
+
+val from_array (#t: Type) (a: A.array t) (#p: perm) (#v: Ghost.erased (Seq.seq t)) : stt (ptr t)
+    (A.pts_to a #p v)
+    (fun s -> exists* fp . pts_to s #p fp v ** is_from_array p fp a)
+
+val to_array (#t: Type) (s: ptr t) (a: array t) (#p: perm) (#fp: footprint t) (#v: Seq.seq t) : stt_ghost unit emp_inames
+    (pts_to s #p fp v ** is_from_array p fp a ** pure (
+        Seq.length v == A.length a
+    ))
+    (fun _ -> A.pts_to a #p v)
+
+(* Written x.(i) *)
+val op_Array_Access
+        (#t: Type)
+        (a: ptr t)
+        (i: SZ.t)
+        (#p: perm)
+        (#fp: footprint t)
+        (#s: Ghost.erased (Seq.seq t))
+  : stt t
+        (requires
+            pts_to a #p fp s ** pure (SZ.v i < Seq.length s))
+        (ensures fun res ->
+            pts_to a #p fp s **
+            pure (
+              SZ.v i < Seq.length s /\
+              res == Seq.index s (SZ.v i)))
+
+(* Written a.(i) <- v *)
+val op_Array_Assignment
+        (#t: Type)
+        (a: ptr t)
+        (i: SZ.t)
+        (v: t)
+        (#fp: footprint t)
+        (#s: Ghost.erased (Seq.seq t))
+  : stt unit
+        (requires
+            pts_to a fp s ** pure (SZ.v i < Seq.length s))
+        (ensures fun res -> exists* s' .
+            pts_to a fp s' **
+            pure (SZ.v i < Seq.length s /\
+              s' == Seq.upd s (SZ.v i) v
+            ))
+
+val share
+  (#a:Type)
+  (arr:ptr a)
+  (#s:Ghost.erased (Seq.seq a))
+  (#fp: footprint a)
+  (#p:perm)
+: stt_ghost unit emp_inames
+      (requires pts_to arr #p fp s)
+      (ensures fun _ -> pts_to arr #(p /. 2.0R) fp s ** pts_to arr #(p /. 2.0R) fp s)
+
+[@@allow_ambiguous]
+val gather
+  (#a:Type)
+  (arr:ptr a)
+  (#s0 #s1:Ghost.erased (Seq.seq a))
+  (#p0 #p1:perm)
+  (#fp: footprint a)
+: stt_ghost unit emp_inames
+      (requires pts_to arr #p0 fp s0 ** pts_to arr #p1 fp s1)
+      (ensures fun _ -> pts_to arr #(p0 +. p1) fp s0 ** pure (s0 == s1))
+
+val adjacent (#t: Type): footprint t -> footprint t -> prop
```

Comment:
I think it would be significantly more ergonomic (and result in a better SMT encoding) to expose base/offset:
```fstar
val base #t : ptr t -> GTot (A.array t)
val offset #t : ptr t -> GTot nat

let adjacent #t (a: ptr t) (sz: nat) (b: ptr t) : prop =
  base a == base b /\ offset a + sz == offset b
```

Then `merge_assoc` becomes unnecessary as well because Z3 can prove it directly.